contracts/RewardsDistributor.sol

```diff
@@ -1,95 +1,369 @@
 
-    function _claimInternal(uint256 index, address account, uint256[] amounts, bytes32[] calldata merkleProof, address actualReceiver) {
-        require(!isClaimed(index), 'RewardsDistributor: Reward already claimed.');
+    /// @notice fires whenever claim() is called
+    event ReferralsClaimed(
+        uint256 period,
+        uint256 index,
+        address account,
+        uint256[] amounts
+    );
 
-        // Verify the merkle proof.
-        bytes32 node = keccak256(abi.encodePacked(index, account, amounts));
-        require(MerkleProof.verify(merkleProof, rewardsRoot, node), 'RewardsDistributor: Invalid proof.');
-
-        // Mark it claimed and send the token.
-        _setClaimed(index);
-        for(uint256 i = 0; i < amounts.length; i++) {
-            if(amounts[i] != 0) {
-                require(IERC20(vaults[i]).transfer(actualReceiver, amounts[i]), 'RewardsDistributor: Transfer failed.');
-            }
-        }
+    /// @notice fires whenever claimMultiple() is called
+    event ReferralsClaimedMultiple(
+        uint256[] periods,
+        uint256[] indexes,
+        address account
+    );
+
+    /// @notice this function checks that the the address is blacklisted, and then calls the _claim() function to send the funds back to the treasury
+    /// @param _period an uin256 that increases every time new rewards are ready to be claimed.
+    /// @param _index a number linked to `msg.sender`
+    /// @param _account the address of the blacklisted account
+    /// @param _amounts an array with the correct amounts of rewards expected
+    /// @param _merkleProof the merkle proof generated using https://www.npmjs.com/package/merkletreejs
+    function claimBlacklist(
+        uint256 _period,
+        uint256 _index,
+        address _account,
+        uint256[] calldata _amounts,
+        bytes32[] calldata _merkleProof
+    ) external onlyRole(BLACKLIST_ROLE) {
+        require(blacklistedAddresses[_account], "Address not blacklisted");
+        _claim(_period, _index, _account, _amounts, _merkleProof, treasury);
+        emit BlacklistClaimed(_period, _index, _account, _amounts, treasury);
     }
 
+    /// @notice this function checks that the the address is not blacklisted, and then calls the _claim() function
+    /// @param _period an uin256 that increases every time new rewards are ready to be claimed.
+    /// @param _index a number linked to `msg.sender`
+    /// @param _amounts an array with the correct amounts of rewards expected
+    /// @param _merkleProof the merkle proof generated using https://www.npmjs.com/package/merkletreejs
+    function claim(
+        uint256 _period,
+        uint256 _index,
+        uint256[] calldata _amounts,
+        bytes32[] calldata _merkleProof
+    ) external {
+        require(!blacklistedAddresses[msg.sender], "Address blacklisted");
+        _claim(_period, _index, msg.sender, _amounts, _merkleProof, msg.sender);
```

Comment:
What happens if we call _claim with a valid Merkle Proof but non corresponding _amounts?